### ISSUE TITLE
InvalidRequestError: Received unknown parameters when updating invoice (from event)

### ISSUE THREAD

**Issue by boydgreenfield** [stripe/stripe-python#505]
**Stripe library version: 2.13.0**

We're having an issue where we'd like to set `auto_advance = False` on invoices meeting certain criteria as they're generated. Unfortunately, we've seen a number of issues with this causing API errors (`InvalidRequestError` and messages about unknown params), and these have proven quite tricky to debug.*

In our backend, we have code that looks something like the following:

```python
def _handle_invoice_create(event=None):
    """
    Automatically disable auto-advance on creation of subscription-linked invoices >$1000 for manual review.
    """
    invoice = event['data']['object']
    if invoice.subscription and invoice.total > 1000 and not invoice.paid:
        invoice.auto_advance = False
        invoice.save()
```

where `event` is being created on our Stripe webhook route with:

```python
event = stripe.Webhook.construct_event(
    payload, received_sig, webhook_secret
)
```

I _believe_ this is a library bug in that: (1) the code used to work with an earlier API version and version `2.0.1` of the client library; and (2) our code is consistent with [the examples in the documentation](https://stripe.com/docs/api/invoices/update). 

It seems like the issue is that the `invoice` object being generated from the event JSON has extra fields that `invoice.save()` is then sending back to the API and they're being rejected. What is the workaround and proper solution here? Do we need to `stripe.Invoice.retrieve(event['data']['object'].id)` and then update the `invoice` object?

Any guidance on either the correct pattern (if this is a "we didn't get the right picture from the documentation" issue) or a workaround (if a bug) would be greatly appreciated.

\* I should note we had the above code working just fine with `invoice.closed = True` back on an earlier Stripe library version and the May 2018 API version, but then an API update broke everything as we started to get error messages about not being able to set both `closed` and `auto_advance`.

**Comment by ob-stripe** [comment]
Hi @boydgreenfield, thanks for the report. Could you share the request ID (`req_...`) from the exception?

**Comment by boydgreenfield** [comment]
@ob-stripe Here are a few: `req_cUcOnZCA8tMwv6`, `req_q6dpdkuhQMpGUT`, `req_q6dpdkuhQMpGUT`

**Comment by ob-stripe** [comment]
@boydgreenfield Thanks! This sounds like a bug with the library. As a workaround, you can try this:
```python
    invoice = event['data']['object']
    if invoice.subscription and invoice.total > 1000 and not invoice.paid:
        stripe.Invoice.modify(invoice.id, auto_advance=False)
```

This should ensure that only the `auto_advance` parameter is sent to the API.

We'll work on a fix and update here when it's ready.

**Comment by boydgreenfield** [comment]
@ob-stripe OK thanks. Is this generally safer syntax to use going forward? Needless to say, it's quite bad when this kind of call fails, so we'd like to take every precaution to ensure that we don't have any issues in the future. Thanks again!

**Comment by ob-stripe** [comment]
@boydgreenfield Yes. `modify` is basically a static version of `save`, it takes the ID of the resource of the resource to update as its first (unnamed) argument, then keyword arguments for each parameter.

It's not very well documented at the moment but it is fully supported, and we're considering making it the canonical way to update resources in the future as it's a lot more explicit than `save` (which does a bunch of internal magic to figure out which attributes have changed and should be sent to the API).

**Comment by boydgreenfield** [comment]
@ob-stripe OK great, thanks!

**Comment by ob-stripe** [comment]
@boydgreenfield I was unable to reproduce the issue you described. I tried something like this:

```python
>>> payload = "..." # JSON string for an invoice.created event object

>>> event = stripe.Event.construct_from(json.loads(payload), stripe.api_key) # this is what stripe.Webhook.construct_event() does, minus the signature verification

>>> type(event)
<class 'stripe.api_resources.event.Event'>

>>> invoice = event['data']['object']

>>> type(invoice)
<class 'stripe.api_resources.invoice.Invoice'>

>>> invoice.serialize(None)
{}

>>> invoice.auto_advance = False

>>> invoice.serialize(None)
{'auto_advance': False}

>>> invoice.save()
<Invoice invoice id=in_1DddVgKCFFPkgtRhyj8oqt6D at 0x112a4e830> JSON: {
...
}
```

It did send the correct request (i.e. with only `auto_advance=False` and no other parameters).

Could you provide more information about what your code does exactly? Something must be messing with the `invoice` instance that causes all the attributes to be sent to the API.

**Comment by ob-stripe** [comment]
Closing due to age, but feel free to reply if this is still an issue.

**Comment by 50gramx** [comment]
I tried creating an Invoice Item with params of charge_automatically, auto_advance and returned the error.

`invoice_item = stripe.InvoiceItem.create(customer=customer_id, price=price_api, auto_advance=True, charge_automatically=True)`

InvalidRequestError:
`stripe.error.InvalidRequestError: Request req_9G7LNMwtpJdKej: Received unknown parameters: auto_advance, charge_automatically`

However, I was able to update these parameters after creating the invoice item without any errors. 

`invoice_item = stripe.InvoiceItem.create(customer=customer_id, price=price_api)`
`invoice_item.auto_advance = True`
`invoice_item.charge_automatically = True`

Posting it here, in case it helps anyone.

Thanks.
